### PR TITLE
feat(uptime): Implement uptime subscription system

### PR DESCRIPTION
**Pull request by wedamija**
This implements a system for creating subscriptions to our url checker. At the moment this is heavily copy/pasted from the snuba subscriptions system, and we plan to abstract the logic between these systems better in the future.

This handles creating and deleting subscriptions. We send subscriptions via kafka, and use tasks to make sure that the subscriptions are successfully created.

Note that we explicitly don't allow `UptimeSubscription` rows to be updated. Since these are shared, we can't modify them without risking affecting other users.

### CODE REVIEW

File: src/sentry/uptime/subscriptions/tasks.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    return {
+        "subscription_id": subscription_id,
+        "url": subscription.url,
+        "interval_seconds": subscription.interval_seconds,  # type: ignore[typeddict-item]
+        "timeout_ms": subscription.timeout_ms,
+    }
+
+
+def send_uptime_config_deletion(subscription_id: str) -> None:
+    send_uptime_config_message(subscription_id, b"")
```
Comment (evanpurkhiser) [inline src/sentry/uptime/subscriptions/tasks.py:128]: It's actually not empty bytes, it's `None`. The types in arroyo are wrong though

File: src/sentry/uptime/subscriptions/tasks.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+def _get_config_codec() -> Codec:
+    global uptime_config_codec
+    if uptime_config_codec is None:
+        uptime_config_codec = get_topic_codec(Topic.UPTIME_CONFIG)
+    return uptime_config_codec
+
+
+def get_uptime_config_topic() -> ArroyoTopic:
+    return ArroyoTopic(get_topic_definition(Topic.UPTIME_CONFIG)["real_topic_name"])
```
Comment (evanpurkhiser) [inline src/sentry/uptime/subscriptions/tasks.py:51]: Can all be replaced with the stuff in `uptime.config_producer`

File: src/sentry/uptime/subscriptions/tasks.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    # Whenever we create/update a config we always want to generate a new subscription id. This allows us to validate
+    # that the config took effect
+    subscription_id = uuid4().hex
+    check_config = uptime_subscription_to_check_config(subscription, subscription_id)
+    send_uptime_config_message(subscription_id, _get_config_codec().encode(check_config))
+    return subscription_id
+
+
+def uptime_subscription_to_check_config(
+    subscription: UptimeSubscription, subscription_id: str
```
Comment (evanpurkhiser) [inline src/sentry/uptime/subscriptions/tasks.py:117]: Why do you have to pass `subscription_id` isn't it in the subscription already?

File: src/sentry/uptime/subscriptions/tasks.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+) -> CheckConfig:
+    return {
+        "subscription_id": subscription_id,
+        "url": subscription.url,
+        "interval_seconds": subscription.interval_seconds,  # type: ignore[typeddict-item]
```
Comment (evanpurkhiser) [inline src/sentry/uptime/subscriptions/tasks.py:122]: Can cast as well if we want to be more explcit.

File: src/sentry/uptime/subscriptions/tasks.py (hunk 1 above)
Comment (evanpurkhiser) [reply src/sentry/uptime/subscriptions/tasks.py:128]: Can be replaced with the stuff in `uptime.config_producer` anyway

File: src/sentry/uptime/subscriptions/tasks.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+
+def send_uptime_config_deletion(subscription_id: str) -> None:
+    send_uptime_config_message(subscription_id, b"")
+
+
+def send_uptime_config_message(subscription_id: str, value: bytes) -> None:
+    producer = _get_subscription_producer()
+    message_key = md5_text(subscription_id).hexdigest().encode()
```
Comment (evanpurkhiser) [inline src/sentry/uptime/subscriptions/tasks.py:133]: In what I wrote the key is just the bytes of the subscription_id. But I think we were talking about making it a composite key of something else?

Right now the uptime checker config consumer expects it to be the subscription key and will fail to consume if it's not https://github.com/getsentry/uptime-checker/blob/eee8326ce849eb11b277a561964e0a2d18835c61/src/config_consumer.rs#L50-L53

**Review by evanpurkhiser — APPROVED** [review]
Looking good.

IMO we should update the SingletonProducer to return the future so we can wait on it, and then we can use that and not have to instantiate the producer ourselves.

There's some differences in the producer you wrote and the one I wrote, so we'll want to reconsicle that.

FYI my producer does get called in a getsentry script, so if we move stuff around we'll want to update that too (though I do think it makes sense to have the producer live in it's own module)

File: src/sentry/uptime/subscriptions/tasks.py (hunk 3 above)
Comment (wedamija) [reply src/sentry/uptime/subscriptions/tasks.py:117]: I don't store it in the subscription until I know I've successfully sent the update to Kafka

File: src/sentry/uptime/subscriptions/tasks.py (hunk 5 above)
Comment (wedamija) [reply src/sentry/uptime/subscriptions/tasks.py:133]: We just need a composite key for the results, not the config

File: src/sentry/uptime/subscriptions/tasks.py (hunk 4 above)
Comment (wedamija) [reply src/sentry/uptime/subscriptions/tasks.py:88]: I remember this being a pain and wasn't sure how to cast to the literal

**Pull request merged**